### PR TITLE
[RFC] vim-patch:7.4.[1730,1734,1741,1742,1779,1782]

### PR DESCRIPTION
**Pull request by brcolow**
Problem:    It is not easy to get a character out of a string.
Solution:   Add strgetchar() and strcharpart().

https://github.com/vim/vim/commit/58de0e2dcc1f2d251b74892a06d71a14973f3187

### CODE REVIEW

**Comment by brcolow** [comment]
Currently stuck on this:

https://travis-ci.org/neovim/neovim/jobs/169981437#L2346

Did I not add the functions in all the right places?

**Comment by jamessan** [comment]
> Did I not add the functions in all the right places?

You did, but you're missing a trailing `FunPtr fptr` argument to each one.  See how the other `f_*` functions have the signature `static void f_name(typval_T *argvars, typval_T *rettv, FunPtr fptr)`?

**Comment by brcolow** [comment]
Thanks @jamessan

**Comment by brcolow** [comment]
Failing tests on Travis:

```
Found errors in Test_strcharpart():
function RunTheTest[9]..Test_strcharpart line 22: Expected '' but got 'x'
function RunTheTest[9]..Test_strcharpart line 23: Expected 'x' but got 'b'
Found errors in Test_strgetchar():
function RunTheTest[9]..Test_strgetchar line 17: Expected 0 but got 120
function RunTheTest[9]..Test_strgetchar line 18: Expected 120 but got 98
```

File: src/nvim/testdir/test_expr_utf8.vim [hunk 1]
```diff
@@ -0,0 +1,26 @@
+" Tests for expressions using utf-8.
+if !has('multi_byte')
+  finish
+endif
+set encoding=utf-8
```
Comment (jamessan) [inline src/nvim/testdir/test_expr_utf8.vim]: This line should be removed.  Changing encoding isn't supported in nvim anymore.

File: src/nvim/testdir/test_expr_utf8.vim [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+scriptencoding utf-8
+
+func Test_strgetchar()
+  call assert_equal(char2nr('á'), strgetchar('áxb', 0))
+  call assert_equal(char2nr('x'), strgetchar('áxb', 1))
+
+  call assert_equal(char2nr('a'), strgetchar('àxb', 0))
+  call assert_equal(char2nr(''), strgetchar('àxb', 1))
```
Comment (jamessan) [inline src/nvim/testdir/test_expr_utf8.vim]: This should be testing that the composing character is returned separately.  You're missing the composing char in the `char2nr` call.

How are you merging the changes?

File: src/nvim/testdir/test_expr_utf8.vim [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  call assert_equal('い', strcharpart('あいうeお', 1, 1))
+  call assert_equal('いう', strcharpart('あいうeお', 1, 2))
+  call assert_equal('いうe', strcharpart('あいうeお', 1, 3))
+  call assert_equal('いうeお', strcharpart('あいうeお', 1, 4))
+  call assert_equal('eお', strcharpart('あいうeお', 3))
+  call assert_equal('e', strcharpart('あいうeお', 3, 1))
+
+  call assert_equal('あ', strcharpart('あいうeお', -3, 4))
+  call assert_equal('a', strcharpart('àxb', 0, 1))
+  call assert_equal('', strcharpart('àxb', 1, 1))
```
Comment (jamessan) [inline src/nvim/testdir/test_expr_utf8.vim]: Same here in regard to the composing character.

**Review by jamessan — CHANGES_REQUESTED** [review]

**Comment by brcolow** [comment]
I currently have on failing assertion:

`function RunTheTest[9]..Test_strcharpart_utf8 line 15: Expected 'a' but got 'à'`

Which is referring to this line:

`call assert_equal('a', strcharpart('àxb', 0, 1))`

Which is in the vim source tree:

https://github.com/vim/vim/blob/master/src/testdir/test_expr_utf8.vim#L34

First of all, I think the more intuitive behavior is that it should return the accented a `à` and not a plain a. So I think the assertion is written incorrectly. For me on Windows gvim running:

`echo strcharpart('àxb', 0, 1)` 

gives me `à` (an accented a).

However Shougo (I assume not on Windows) gets a plain `a`.

So questions:

* Why are we getting  `à` instead of `a`?
* Should we just change the assertion to expect `à` instead of `a` as it is more intuitive?
* Is this somehow related to the fact that neovim only supports utf8?

**Comment by justinmk** [comment]
Examining the raw file: 

```
    curl -LO https://raw.githubusercontent.com/vim/vim/master/src/testdir/test_expr_utf8.vim
    nvim test_expr_utf8.vim
```

Inspecting `á` in this line with the `ga` command (enhanced by [characterize.vim](https://github.com/tpope/vim-characterize):

    call assert_equal('á', strcharpart('áxb', 0, 1))
    <á> 225, \341, U+00E1 LATIN SMALL LETTER A WITH ACUTE, ^Ka', &aacute; 

then comparing `à` in this line:

    call assert_equal('a', strcharpart('àxb', 0, 1))
    <a> 97, \141, U+0061 LATIN SMALL LETTER A < ̀> 768, U+0300 COMBINING GRAVE ACCENT

we can see that the `à` is two chars: `a` plus a combining character. Whereas `á` in the previous case is a single-byte ASCII character. That's the distinction being tested in the tests.

> For me on Windows gvim running: `echo strcharpart('àxb', 0, 1)` gives me à (an accented a).

Hmm, that's not what we want.

**Comment by jamessan** [comment]
> Why are we getting à instead of a?

Because the test file has the single Unicode codepoint for à rather than the a and the combining grave accent.

Every test using á is using the single Unicode codepoint and every test using à is using a + the combining grave.  This is explicitly to test the handling of combining characters.

File: src/nvim/testdir/test_expr_utf8.vim [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  call assert_equal('い', strcharpart('あいうeお', 1, 1))
+  call assert_equal('いう', strcharpart('あいうeお', 1, 2))
+  call assert_equal('いうe', strcharpart('あいうeお', 1, 3))
+  call assert_equal('いうeお', strcharpart('あいうeお', 1, 4))
+  call assert_equal('eお', strcharpart('あいうeお', 3))
+  call assert_equal('e', strcharpart('あいうeお', 3, 1))
+
+  call assert_equal('あ', strcharpart('あいうeお', -3, 4))
+
+  call assert_equal('a', strcharpart('àxb', 0, 1))
```
Comment (jamessan) [inline src/nvim/testdir/test_expr_utf8.vim]: This line isn't using the combining grave accent.  The other two tests below are.

**Comment by brcolow** [comment]
I think I addressed all of your comments @jamessan . Also, tests are now passing.

**Comment by Shougo** [comment]
Nice.

**Review by jamessan — APPROVED** [review]

**Pull request merged**